pyhanko/sign/signers/csc_signer.py

```diff
@@ -0,0 +1,155 @@
+
+The 2 points marked with ! are CSC calls and are considered external to this module.
+
+- prepare_document(pdf) => digest
+! get_certificates_from_CSC_server() => cert_chain
+- build_attributes_to_be_signed(digest, cert_chain) => cms_attributes
+! sign_with_CSC_server(cms_attributes) => signature_pkcs1
+- generate_cms(digest, cert_chain, signature_pkcs1) => signed_cms_attributes
+- attach_signed_CMS_to_document(pdf, signed_cms_attributes) => signed_pdf
+"""
+from base64 import b64decode
+from io import BytesIO
+from typing import Any, TypedDict
+
+from pyhanko.sign import signers, timestamps, fields
+from pyhanko.sign.general import SimpleCertificateStore
+from pyhanko.sign.signers.pdf_signer import PdfTBSDocument
+from pyhanko.pdf_utils.incremental_writer import IncrementalPdfFileWriter
+from pyhanko_certvalidator import ValidationContext
+from asn1crypto import x509
+
+
+SigSeedSubFilter = fields.SigSeedSubFilter
+
+
+def as_any(_) -> Any:
+    return _
+
+
+class TDigest(TypedDict):
+    data_digest: bytes
+    digest_algorithm: str
+
+
+class TPrepDocumentConfig(TypedDict, total=False):
+    embed_validation_info: bool
+    subfilter: SigSeedSubFilter
+    validation_context: ValidationContext
+    md_algorithm: str
+
+
+class CscSigner(signers.ExternalSigner):
+    """
+    See https://cloudsignatureconsortium.org/resources/download-api-specifications/
+    """
+
+    def __init__(
+        self,
+        csc_credentials_info_response: dict,
+        tsa_url: str,
+        signature_value=bytes(32),
+        use_pades=False,
+    ):
+        """
+        csc_credentials_info_response:
+            List returned by CSC credentials/info endpoint with "certificates"
+            request parameter set to "chain". See section 11.5 of the CSC spec.
+        """
+        self.use_pades = use_pades
+        self.timestamper = timestamps.HTTPTimeStamper(tsa_url)
+        certs_encoded = csc_credentials_info_response['cert/certificates']
+        certs = [x509.Certificate.load(b64decode(cert)) for cert in certs_encoded]
+        cert_registry = SimpleCertificateStore()
+        cert_registry.register_multiple(set(certs))
+        super().__init__(
+            signing_cert=certs[0],
+            cert_registry=cert_registry,
+            signature_mechanism=None,
+            signature_value=signature_value,
+        )
+
+    def prep_document(self, pdf: bytes, config: TPrepDocumentConfig):
+        """
+        This method is pdf specific.
+        It returns a pdf document that is ready to have a CMS attached.
+        Also it computes the pdf specific hash of the prepared document.
+        """
+        pdf_signer = signers.PdfSigner(
+            signers.PdfSignatureMetadata(
+                field_name='SigNew',
+                embed_validation_info=config.get("embed_validation_info", False),
+                use_pades_lta=self.use_pades,
+                subfilter=config.get("subfilter", SigSeedSubFilter.ADOBE_PKCS7_DETACHED),
```

Comment:
Yes, I should probably do a better job of explaining that. The idea is that subfilters represent a set of conventions around the format of the signature, so that different "filters" (i.e. signature implementations) can still interoperate with one another. The "default" is `/adbe.pkcs7.detached`, which sort of serves as a catch-all (it's more complicated than that, but that'll do as a first-order approximation). For PAdES signatures, you're supposed to use `/ETSI.CAdES.detached`.

In particular, using `use_pades_lta` with `/adbe.pkcs7.detached` doesn't really make sense (and I believe that is enforced in the library code, but I'm not 100% sure).

For what it's worth, these things being called "filters" and "subfilters" is largely a historical accident. There's no deeper meaning behind that terminology, other than that it's what's used in the PDF standard.

---------

Anyway, to my original question: would you be OK with allowing the `PdfSignatureMetadata` to be passed in as an optional parameter? You can still provide a default if you want to.

Reply:
Removed signature_meta creation.